File: MP.Web/ClientApp/src/app/services/tag/tag.service.ts
import { Injectable, Inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { BASE_URL } from '@mintplayer/ng-base-url';
import { PaginationRequest, PaginationResponse } from '@mintplayer/ng-pagination';
import { Tag } from '../../entities/tag';

@Injectable({
  providedIn: 'root'
})
export class TagService {
  constructor(private httpClient: HttpClient, @Inject(BASE_URL) private baseUrl: string, @Inject('API_VERSION') private apiVersion: string) {
  }

  public pageTags(request: PaginationRequest) {
    return this.httpClient.post<PaginationResponse<Tag>>(`${this.baseUrl}/web/${this.apiVersion}/Tag/page`, request).toPromise();
  }

  public suggestTags(search: string) {
    return this.httpClient.post<Tag[]>(`${this.baseUrl}/web/${this.apiVersion}/Tag/suggest`, { searchTerm: search }).toPromise();
  }

  public searchTags(search: string) {
    return this.httpClient.post<Tag[]>(`${this.baseUrl}/web/${this.apiVersion}/Tag/search`, { searchTerm: search }).toPromise();
  }

  public getTags(include_relations: boolean) {
    return this.httpClient.get<Tag[]>(`${this.baseUrl}/web/${this.apiVersion}/Tag`, {
      headers: {
        'include_relations': String(include_relations)
      }
    }).toPromise();
  }

  public getTag(id: number, include_relations: boolean) {
    return this.httpClient.get<Tag>(`${this.baseUrl}/web/${this.apiVersion}/Tag/${id}`, {
      headers: {
        'include_relations': String(include_relations)
      }
    }).toPromise();
  }

  public createTag(tag: Tag) {
    let clone = this.removeSubjects(tag);
    return this.httpClient.post<Tag>(`${this.baseUrl}/web/${this.apiVersion}/Tag`, clone).toPromise();
  }

  public updateTag(tag: Tag) {
    let clone = this.removeSubjects(tag);
    return this.httpClient.put<Tag>(`${this.baseUrl}/web/${this.apiVersion}/Tag/${tag.id}`, clone).toPromise();
  }

  private removeSubjects(tag: Tag) {
    // Remove "subjects" from tag
    let clone = Object.assign({}, tag);
    clone.subjects = [];
    if (clone.parent !== null) {
      clone.parent.subjects = [];
    }

    return clone;
  }

  public deleteTag(tag: Tag) {
    return this.httpClient.delete(`${this.baseUrl}/web/${this.apiVersion}/Tag/${tag.id}`).toPromise();
  }
}

File: legacy/MintPlayer.Data/Services/TagService.cs
using MintPlayer.Dtos.Dtos;
using MintPlayer.Data.Repositories;
using MintPlayer.Pagination;
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading.Tasks;
using MintPlayer.Data.Abstractions.Services;

namespace MintPlayer.Data.Services
{
    internal class TagService : ITagService
    {
        private readonly ITagRepository tagRepository;
        public TagService(ITagRepository tagRepository)
        {
            this.tagRepository = tagRepository;
        }

        public async Task<PaginationResponse<Tag>> PageTags(PaginationRequest<Tag> request)
        {
            var tags = await tagRepository.PageTags(request);
            return tags;
        }

        public async Task<IEnumerable<Tag>> Suggest(string search_term, bool include_relations = false)
        {
            var tags = await tagRepository.Suggest(search_term, include_relations);
            return tags;
        }

        public async Task<IEnumerable<Tag>> Search(string search_term)
        {
            var tags = await tagRepository.Search(search_term);
            return tags;
        }

        public async Task<IEnumerable<Tag>> GetTags(bool root_tags_only = true, bool include_relations = false)
        {
            var tags = await tagRepository.GetTags(root_tags_only, include_relations);
            return tags;
        }

        public async Task<Tag> GetTag(int id, bool include_relations = false)
        {
            var tag = await tagRepository.GetTag(id, include_relations);
            return tag;
        }

        public async Task<Tag> InsertTag(Tag tag)
        {
            var newTag = await tagRepository.InsertTag(tag);
            return newTag;
        }

        public async Task<Tag> UpdateTag(Tag tag)
        {
            var updatedTag = await tagRepository.UpdateTag(tag);
            await tagRepository.SaveChangesAsync();
            return updatedTag;
        }

        public async Task DeleteTag(int tag_id)
        {
            await tagRepository.DeleteTag(tag_id);
            await tagRepository.SaveChangesAsync();
        }
    }
}
